ledger/accountdb.go

```diff
@@ -1131,3 +1131,232 @@ func (iterator *encodedAccountsBatchIter) Close() {
+	oaiStepDeleteOldOrderingTable = orderedAccountsIterStep(0)
+	// create new ordering table
+	oaiStepCreateOrderingTable = orderedAccountsIterStep(1)
+	// query the existing accounts
+	oaiStepQueryAccounts = orderedAccountsIterStep(2)
+	// iterate over the existing accounts and insert their hash & address into the staging ordering table
+	oaiStepInsertAccountData = orderedAccountsIterStep(3)
+	// create an index on the ordering table so that we can efficiently scan it.
+	oaiStepCreateOrderingAccountIndex = orderedAccountsIterStep(4)
+	// query the ordering table
+	oaiStepSelectFromOrderedTable = orderedAccountsIterStep(5)
+	// iterate over the ordering table
+	oaiStepIterateOverOrderedTable = orderedAccountsIterStep(6)
+	// cleanup and delete ordering table
+	oaiStepShutdown = orderedAccountsIterStep(7)
+	// do nothing as we're done.
+	oaiStepDone = orderedAccountsIterStep(8)
+)
+
+// orderedAccountsIter allows us to iterate over the accounts addresses in the order of the account hashes.
+type orderedAccountsIter struct {
+	step             orderedAccountsIterStep
+	rows             *sql.Rows
+	tx               *sql.Tx
+	accountCount     int
+	fetchAccountData bool
+	insertStmt       *sql.Stmt
+}
+
+// makeOrderedAccountsIter creates an ordered account iterator. Note that due to implementation reasons,
+// only a single iterator can be active at a time.
+func makeOrderedAccountsIter(tx *sql.Tx, accountCount int, fetchAccountData bool) *orderedAccountsIter {
+	return &orderedAccountsIter{
+		tx:               tx,
+		accountCount:     accountCount,
+		fetchAccountData: fetchAccountData,
+		step:             oaiStepStartup,
+	}
+}
+
+// accountAddressHashData is used by Next to return a single account address, hash and account data.
+type accountAddressHashData struct {
+	address            basics.Address
+	digest             []byte
+	encodedAccountData []byte
+}
+
+// Next returns an array containing the account address, hash and potentially data
+// the Next function works in multiple processing stages, where it first processs the current accounts and order them
+// followed by returning the ordered accounts. In the first phase, it would return empty accountAddressHashData array
+// whose size matches with the data that was processed, and set the ordered to false. On the second phase, the acct
+// would contain valid data ( and optionally the account data as well, if was asked in makeOrderedAccountsIter) and
+// would set the ordered to true. If err is sql.ErrNoRows it means that the iterator have completed it's work and no further
+// accounts exists. Otherwise, the caller is expected to keep calling "Next" to retrieve the next set of accounts
+// ( or let the Next function make some progress toward that goal )
+func (iterator *orderedAccountsIter) Next(ctx context.Context) (acct []accountAddressHashData, ordered bool, err error) {
+	if iterator.step == oaiStepDeleteOldOrderingTable {
```

Comment:
The breaking into stages is designed so that the entire process can be paused at any moment. To accomplish that, I broke that into granular sql statements.